### PR TITLE
Limit digilines

### PR DESCRIPTION
**Pull request by numberZero**
Modified #379 
@Hawk777 @20kdc what’s your opinions on this variant?

### CODE REVIEW

**Comment by 20kdc** [comment]
Something like this got committed to master by Jejia over a year ago.
See: https://github.com/minetest-mods/mesecons/commit/703e6fdadb5251b6f42e35f0f71f3094f5e15f75 and discussion there.
Then this happened... https://github.com/minetest-mods/mesecons/commit/54daee236e680fa79b2dd31e812f7e063887a0bf

**Comment by 20kdc** [comment]
...To clarify on the previous, I'm not saying this is a bad approach for this particular issue. I *am* saying I don't know if it'll stick around, given what happened last time this approach was chosen. That said, the previous approach used both serialize and something like this - perhaps this will be faster than serialize.

**Comment by numberZero** [comment]
But I *replace* serialization test with a special function, so this should be actually even more performant than the original code.

**Comment by numberZero** [comment]
Testing needed! Preferably with something as heavy as LTC-4000, does anyone have experience with it?

**Comment by Hawk777** [comment]
At first glance this looks like a pure rewrite of #379, so what’s the difference?

**Comment by 20kdc** [comment]
I went and read through both, seems to me like it's based on PR 379 but with additions.

**Comment by numberZero** [comment]
@Hawk777
1. It refuses to send anything containing disallowed data, instead of “cleaning up” an invalid message.
2. It returns early if the message appears to be too heavy.

File: mesecons_luacontroller/init.lua [hunk 1]
```diff
@@ -269,28 +269,60 @@ local function get_interrupt(pos)
...
+		return msg, 8
+	elseif t == "boolean" then
+		return msg, 1
+	elseif t == "table" then
+		local bref = back_references[msg]
+		if bref then
+			return bref, 0
+		end
+		local cost = 8
+		local ret = {}
```
Comment (Hawk777) [inline mesecons_luacontroller/init.lua]: You never actually put anything in `ret`, do you?

File: mesecons_luacontroller/init.lua [hunk 2]
```diff
@@ -269,28 +269,60 @@ local function get_interrupt(pos)
 	return interrupt
 end
 
+local function safe_deep_copy_helper(msg, back_references, size_limit)
+	local t = type(msg)
+	if t == "string" then
+		return msg, #msg + 25
```
Comment (Hawk777) [inline mesecons_luacontroller/init.lua]: A long string won’t be checked against `size_limit`.

**Comment by Hawk777** [comment]
Meh. Seems like this is exactly identical to #379 if you were to say yes to my question at https://github.com/minetest-mods/mesecons/pull/379#discussion_r161133901, plus duplicating received tables. If you think it’s easier to make a whole new PR rather than just edit the existing one, go ahead.

File: mesecons_luacontroller/init.lua (hunk 1 above)
Comment (numberZero) [reply mesecons_luacontroller/init.lua]: er, my mistake.

File: mesecons_luacontroller/init.lua (hunk 2 above)
Comment (numberZero) [reply mesecons_luacontroller/init.lua]: it will. see `safe_deep_copy`

**Comment by numberZero** [comment]
@Hawk777 Sorry for inconvenience, it’s easier to copy a PR and edit the copy, than to edit the PR branch directly.

File: mesecons_luacontroller/init.lua (hunk 1 above)
Comment (numberZero) [reply mesecons_luacontroller/init.lua]: fixed.

**Comment by Hawk777** [comment]
… but I was editing the branch all along, fixing your comments, without any trouble. Oh well.

File: mesecons_luacontroller/init.lua (hunk 2 above)
Comment (Hawk777) [reply mesecons_luacontroller/init.lua]: Yes it will, I see now the size limit is checked in two places. Sorry about that.

**Comment by numberZero** [comment]
Merged #379.

**Pull request closed without merging**